transifex/common/utils.py

```diff
@@ -6,14 +6,33 @@
 import pytz
 
 
-def generate_key(source_string, context=None):
+def generate_key(string, context=None, parsed=False, plurals=None):
     """Return a unique key based on the given source string and context.
 
-    :param unicode source_string: the source string
-    :param Union[unicode, list] context: an optional context that accompanies the string
+    :param str string: An ICU string
+    :param dict plurals: A dictionary with pre-parsed plurals.
+        Should be like:
+            { 1: "Here is one dog",
+              5: "Here are many dogs"}
+    :param Union[unicode, list] context: an optional context that
+        accompanies the string
     :return: a unique key
     :rtype: str
     """
+    if not string and not plurals:
+        raise ValueError("You need to specify at least "
+                         "one of `string`, `plurals`")
+    if string and plurals:
+        raise ValueError("Cannot use both `string` and `plurals`")
+
+    if not plurals:
+        _, plurals = parse_plurals(string)
+
+    string_content = u':'.join(
+        u'{}:{}'.format(rule, string)
+        for rule, string in sorted(plurals.items())
+    )
```

Comment:
1. Is sorting `.items()` guaranteed to be consistent?
2. if we split with `:`, I think we need to escape `:`s in the strings themselves:

   ```python
   strings1 = {1: "a:5", 5: "b"}
   strings2 = {1: "a", 5: "5:b"}
   
   # Without escaping
   string_content1 = "1:a:5:5:b"
   string_content2 = "1:a:5:5:b"
   
   # With escaping
   string_content1 = "1:a\:5:5:b"
   string_content2 = "1:a:5:5\:b"
   ```

   (and then we would also have to escape the `\`s in the strings)

Reply:
1. Since it's a list of tuples, I think it should be.
2. True, didn't think of this.